File: LC146.cpp
//o(1) 复杂度，需要用map来登记key和vector的关系
class LRUCache {
private:
    vector<pair<int, int>> data;
    int capacity;
    unordered_map<int, vector<pair<int, int>>::iterator> m;
public:
    LRUCache(int capacity) {
        this->capacity = capacity;
    }
    
    int get(int key) {
        bool flag_found = false;
        int res;
        
        auto it=m.find(key);
        if(it==m.end()) return -1;

        res = (*it).second->second;
        auto temp_val = (*it).second->second;
        this->data.erase(it->second); //deque
        // this->data.push_back(temp_val); //inque
        
        return res;
                
    }
    
    void put(int key, int value) {
        auto it=m.find(key);
        if(it==m.end()){
            this->data.push_back({key,value}); //inque
            // m[key] = data.rbegin();
        }else{
            // this->data.erase(it);
            this->data.push_back({key,value}); //inque
            // this->m[key]=this->data.rbegin();
        }    
        if(this->data.size()>=this->capacity){
          this->data.erase(data.begin());  
        } 
    }
};

/**
 * Your LRUCache object will be instantiated and called as such:
 * LRUCache* obj = new LRUCache(capacity);
 * int param_1 = obj->get(key);
 * obj->put(key,value);
 */

File: 1st/LC146.py
class LRUCache(object):

    def __init__(self, capacity):
        """
        :type capacity: int
        """
        self.cash = collections.defaultdict(int)
        self.cap = capacity
        self.order = [] # 记录各个key进来的oder，越后面越新
        

    def get(self, key):
        """
        :type key: int
        :rtype: int
        """
        # 要把用过的最新的放order最后面
        if key in self.order:
            self.order.remove(key)
            self.order =  self.order + [key]
        
        return self.cash[key]-1
        pass
        

    def put(self, key, value):
        """
        :type key: int
        :type value: int
        :rtype: None
        """
        
      
        
        
        if len(self.order) >= self.cap and not key in self.order:
            self.cash[self.order[0]] = 0
            self.order = self.order[1:]
            
        # 要把用过的最新的放order最后面
        if key in self.order:
            self.order.remove(key)
            self.order =  self.order + [key]  
        else:
            self.order.append(key)
            
        self.cash[key] = value+1
        
        
        


# Your LRUCache object will be instantiated and called as such:
# obj = LRUCache(capacity)
# param_1 = obj.get(key)
# obj.put(key,value)